cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -428,183 +408,152 @@ func getAPIGroupPreferredVersion(client discovery.DiscoveryInterface, APIGroup s
-	}
-	for _, machineSet := range machineSets {
-		if err := f(machineSet); err != nil {
-			return err
+		if found && nodeRefKind != "Node" {
+			klog.Errorf("Status.NodeRef of machine %q does not reference a node (rather %q)", machine.GetName(), nodeRefKind)
+			continue
 		}
-	}
-	return nil
-}
-
-func (c *machineController) machineSetNodeGroups() ([]*nodegroup, error) {
-	var nodegroups []*nodegroup
 
-	if err := c.filterAllMachineSets(func(machineSet *MachineSet) error {
-		if machineSetHasMachineDeploymentOwnerRef(machineSet) {
-			return nil
-		}
-		ng, err := newNodegroupFromMachineSet(c, machineSet)
+		nodeRefName, found, err := unstructured.NestedString(machine.Object, "status", "nodeRef", "name")
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		if ng.MaxSize()-ng.MinSize() > 0 {
-			if ng.scalableResource.CanScaleFromZero() {
-				nodegroups = append(nodegroups, ng)
-			} else if pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, machineSet.Status.Replicas) > 0 {
-				nodegroups = append(nodegroups, ng)
+		if found {
+			node, err := c.findNodeByNodeName(nodeRefName)
+			if err != nil {
+				return nil, fmt.Errorf("unknown node %q", nodeRefName)
+			}
+
+			if node != nil {
+				providerIDs = append(providerIDs, node.Spec.ProviderID)
 			}
 		}
-		return nil
-	}); err != nil {
-		return nil, err
 	}
 
-	return nodegroups, nil
+	klog.V(4).Infof("nodegroup %s has nodes %v", scalableResource.GetName(), providerIDs)
+
+	return providerIDs, nil
 }
 
-func (c *machineController) machineDeploymentNodeGroups() ([]*nodegroup, error) {
-	machineDeployments, err := c.listMachineDeployments(metav1.NamespaceAll, labels.Everything())
+func (c *machineController) nodeGroups() ([]*nodegroup, error) {
+	scalableResources, err := c.listScalableResources()
 	if err != nil {
 		return nil, err
 	}
 
-	var nodegroups []*nodegroup
+	nodegroups := make([]*nodegroup, 0, len(scalableResources))
 
-	for _, md := range machineDeployments {
-		ng, err := newNodegroupFromMachineDeployment(c, md)
+	for _, r := range scalableResources {
+		ng, err := newNodegroupFromScalableResource(c, r)
 		if err != nil {
 			return nil, err
 		}
+
 		// add nodegroup iff it has the capacity to scale
 		if ng.MaxSize()-ng.MinSize() > 0 {
+			// if the node group can scale from zero there is no need to check the replicas against 0, it should be added
 			if ng.scalableResource.CanScaleFromZero() {
 				nodegroups = append(nodegroups, ng)
-			} else if pointer.Int32PtrDerefOr(md.Spec.Replicas, 0) > 0 {
-				nodegroups = append(nodegroups, ng)
+				continue
 			}
-		}
-	}
 
-	return nodegroups, nil
-}
+			var replicas int64
+			replicas = 0
 
-func (c *machineController) nodeGroups() ([]*nodegroup, error) {
-	machineSets, err := c.machineSetNodeGroups()
-	if err != nil {
-		return nil, err
-	}
+			// to avoid a situation where spec.replicas is nil, we first get the value and inspect it
+			value, found, err := unstructured.NestedFieldNoCopy(r.Object, "spec", "replicas")
```

Comment:
i _think_ that should work, i have a feeling i was following the changeset here but i can certainly move this logic into the Replicas() call.

Reply:
i have reverted this code back to something like the original. since we are using the scale resource we should never reach a state where we have the nil pointer here.